### ISSUE TITLE
Restrict hyperparameter choice based on data format

### ISSUE THREAD

**Issue by bjkomer** [hyperopt/hyperopt-sklearn#28]
For example, KNN has a distance metric parameter and some metrics cannot be used on sparse data (e.g. chebyshev). Need a nice way to prevent these from being selected when sparse data is used.

One way could be to have a separate search space defined for sparse and dense data. (i.e. knn() and knn_sparse())
Another option could be to have a sparse/dense flag that changes how the space is defined (i.e. knn( sparse=True ))

I'm leaning towards the second option.
